<a href="https://colab.research.google.com/github/HadeerMahdy19/task1/blob/main/Find_technical_skills.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import requests
import nltk
import pandas as pd
import numpy as np

### steps:
- lower strings
- drop duplicates
- filter skills from puctuation and stop words




In [2]:
data_url = 'https://raw.githubusercontent.com/HadeerMahdy19/task1/main/Raw_Skills_Dataset.csv'

In [3]:
raw_df = pd.read_csv(data_url, encoding='unicode_escape')
raw_df.tail(10)

RAW DATA
34106                           Leadership qualities
34107                            limited supervision
34108                                       Actalent
34109                                 all applicants
34110  Self-motivated, enthusiastic and strong drive
34111                                    negotiation
34112                                      deadlines
34113  Self-motivated, enthusiastic and strong drive
34114                                    negotiation
34115                                      deadlines

In [4]:
raw_df.shape

(34116, 1)

### Remove punctuation and stop words

In [5]:
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

In [6]:
nltk.download('punkt')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [7]:
stop_words = set(nltk.corpus.stopwords.words('english'))


--------------------------------------------------------------------------

In [8]:
def preprocess_data(data):
  # convert all skills to lower case 
  data_lower = data.iloc[:,0].apply(lambda s: s.lower())

  # drop duplicates
  data_lower.drop_duplicates(inplace = True)
  print(f"Data size after lowering capitals + dropping duplicates : {data_lower.shape}")

  # extract tokens from each skill
  tokenized_df = data_lower.apply(lambda s: nltk.tokenize.word_tokenize(s))

  preprocessed_df = tokenized_df.copy()

  # filter each skill from stop words and puctuations
  for i in tokenized_df.index: 
    filtered_skill = []    # store one skill tokens after cleaning

    for skill_token in tokenized_df[i]:
      if skill_token not in stop_words and skill_token.isalpha():
        filtered_skill.append(skill_token)
    preprocessed_df[i] = ' '.join(filtered_skill)  # make filtered skill tokens into one sentence

  # drop duplicates after filtering
  preprocessed_df.drop_duplicates(inplace = True)
  print(f"Data size after tokenization + dropping stop words & punct + dropping duplicates : {preprocessed_df.shape}")
 
  return preprocessed_df


- Process and clean our skills df 

In [9]:
preprocessed_df = preprocess_data(raw_df)

Data size after lowering capitals + dropping duplicates : (14892,)
Data size after tokenization + dropping stop words & punct + dropping duplicates : (12725,)


- Process and clean technical skills df from emsi API

In [10]:
technical_skills_df = pd.read_csv('https://raw.githubusercontent.com/HadeerMahdy19/task1/main/technical_skills.csv', encoding='unicode_escape')

In [11]:
preprocessed_technical_skills_df = preprocess_data(technical_skills_df.drop(['Unnamed: 0'],axis=1))

Data size after lowering capitals + dropping duplicates : (31993,)
Data size after tokenization + dropping stop words & punct + dropping duplicates : (30736,)


- Find similar skills from preprocessed_df in preprocessed_technical_skills_df

In [12]:
res = []
for s in preprocessed_df:
  if np.sum(preprocessed_technical_skills_df.str.contains(s, case=False)) >= 1:
    res.append(s)


In [13]:
res 

['ifs',
 'relational databases',
 'sql',
 'orm',
 '',
 'hibernate',
 'mybatis',
 'maven',
 'continuous integration',
 'jenkins',
 'bamboo',
 'frameworks',
 'transformations',
 'consistency',
 'methodologies',
 'data',
 'phases',
 'data warehouse',
 'python',
 'postgres',
 'mysql',
 'oracle',
 'clickhouse',
 'apache spark',
 'apache flink',
 'apache kafka',
 'aws kinesis',
 'gcp',
 'azure',
 'aws',
 'persuasion',
 'scripting',
 'kafka',
 'grpc',
 'docker',
 'kubernetes',
 'cloud environments',
 'etcd',
 'vault',
 'load balancers',
 'databases',
 'raft',
 'paxos',
 'github',
 'microservices',
 'ruby',
 'bazel',
 'http',
 'redis',
 'javascript',
 'html',
 'css',
 'algorithms',
 'angular',
 'meteor',
 'ember',
 'knockout',
 'vue',
 'bootstrap',
 'typescript',
 'commonjs',
 'stylesheet languages',
 'sass',
 'redux',
 'rxjs',
 'apis',
 'json',
 'architectural patterns',
 'proficiency',
 'concepts',
 'latency',
 'heap',
 'profiling',
 'nosql',
 'mongodb',
 'cassandra',
 'oracle weblogic',
 'j

In [14]:
len(res)

2338

## - `res` is the list of final result from preprocessing and finding our technical skills

### Technical skills found in Raw_Skills_Dataset.csv are (2338) skill